typescript/sdk/src/token/read.ts

```diff
@@ -44,34 +66,63 @@ export class EvmERC20WarpRouteReader {
    */
   async deriveWarpRouteConfig(
     address: Address,
-  ): Promise<DerivedERC20WarpRouteConfig> {
+  ): Promise<DerivedTokenRouterConfig> {
     const fetchedBaseMetadata = await this.fetchBaseMetadata(address);
     const fetchedTokenMetadata = await this.fetchTokenMetadata(
       fetchedBaseMetadata.token,
     );
 
-    const results: DerivedERC20WarpRouteConfig = {
+    // Derive the config type
+    const type = await this.deriveTokenType(address);
+
+    const derivedTokenRouter: DerivedTokenRouterConfig = {
       ...fetchedBaseMetadata,
       ...fetchedTokenMetadata,
+      type,
     };
 
-    if (
-      fetchedBaseMetadata.interchainSecurityModule !==
-      ethers.constants.AddressZero
-    ) {
-      results.interchainSecurityModule =
-        await this.evmIsmReader.deriveIsmConfig(
-          fetchedBaseMetadata.interchainSecurityModule,
+    return derivedTokenRouter;
+  }
+
+  /**
+   * Derives the token type for a given Warp Route address using specific methods
+   *
+   * @param address - The Warp Route address to derive the token type for.
+   * @returns The derived token type, which can be one of: collateralVault, collateral, native, or synthetic.
+   */
+  async deriveTokenType(address: Address): Promise<DerivedTokenType> {
+    const contractTypes: Record<
+      Exclude<DerivedTokenType, 'native'>, // native is excluded because it's the default return type
+      { factory: any; method: string }
+    > = {
+      collateral: {
+        factory: HypERC20Collateral__factory,
+        method: 'wrappedToken',
+      },
+      collateralVault: {
+        factory: HypERC20CollateralVaultDeposit__factory,
+        method: 'vault',
+      },
+      synthetic: {
+        factory: HypERC20__factory,
+        method: 'decimals',
+      },
+    };
+
+    for (const [type, { factory, method }] of Object.entries(contractTypes)) {
+      try {
+        const warpRoute = factory.connect(address, this.provider);
+        await warpRoute[method]();
+        return type as DerivedTokenType;
+      } catch (e) {
+        this.logger.debug(
+          `Error accessing token specific property, implying this is not a ${type} token. Defaulting to ${TokenType.native}.`,
+          address,
         );
+      }
```

Comment:
Maybe this is the best we can do but IMO relying on an exception as part of expected control flow is a code smell. Wondering if there are any alternatives we can consider?

Reply:
Agree, but unfortunately, we don't have a `type` on the bytecode so this is the best we can do, for now.

@yorhodes creating a ticket for this